okhttp/src/commonTest/kotlin/okhttp3/HttpUrlTest.kt

```diff
@@ -1838,4 +1839,23 @@ open class HttpUrlTest {
       .isEqualTo("http://a$dotA126/")
     assertInvalid("http://aa$dotA126/", "Invalid URL host: \"aa$dotA126\"")
   }
+
+  /**
+   * UTS 46 Validity Criteria: Decoded punycode must be NFC.
+   *
+   * https://www.unicode.org/reports/tr46/#Validity_Criteria
+   */
+  @Test
+  fun hostnameInPunycodeNfcAndNfd() {
+    // café can be NFC (é is one code point) or NFD (e plus ´ as two code points).
+    val hostNfc = "café.com"
+    val hostNfcPunycode = "xn--caf-dma.com"
+    val hostNfd = "café.com"
+    val hostNfdPunycode = "xn--cafe-yvc.com"
+    assertEquals(hostNfcPunycode, "http://$hostNfc/".toHttpUrl().host)
+    assertEquals(hostNfcPunycode, "http://$hostNfcPunycode/".toHttpUrl().host)
+    assertEquals(hostNfcPunycode, "http://$hostNfd/".toHttpUrl().host)
+    if (isJvm) return // TODO: the rest of this test is broken on JVM platforms.
+    assertInvalid("http://$hostNfdPunycode/", """Invalid URL host: "$hostNfdPunycode"""")
```

Comment:
java.net.IDN won’t decode hostNfdPunycode from punycode (presumably 'cause it’s NFD), but it will decode hostNfcPunycode. So the behavior is kind of there.

But java.net.URL handles both.